### DISCUSSION TITLE
Problems with mcrun.

### DISCUSSION THREAD

**Discussion by Frida854** [Moddable-OpenSource/moddable#715] [Q&A] [answered]
Got a new mac pro with m1 processor.
Everything is up and running now.
In ch1-getting started,
have I run helloworld-gui, works.
have I run host, works.
But helloworld gives me problems.

When I run mcconfig ..., I see the following lines:
```
Executing action: flash
Serial port /dev/cu.wlan-debug
/dev/cu.wlan-debug failed to connect: [Errno 16] could not open port /dev/cu.wlan-debug: [Errno 16] Resource busy: '/dev/cu.wlan-debug'
Serial port /dev/cu.usbserial-0001
Connecting ........_____....._____.
```
But when I run mcrun ...., I see the following lines:
```
poulborg @ Poul-MacBook-Pro helloworld% mcrun -d -m -p esp32 / moddable_two
# xsc example.xsb
# xsc check.xsb
# xsc mod / config.xsb
# xsl helloworld.xsa
Error opening serial port /dev/cu.wlan-debug - Resource busy (16).
```
mcrun, does not switch on automatically
```
Serial port /dev/cu.usbserial-0001
```
as mcconfig does.

By changing a little in
make.esp32.mk
```
ifeq ($ (HOST_OS), Darwin)
UPLOAD_PORT = /dev/cu.usbserial-0001
endif


#ifeq ($ (HOST_OS), Darwin)
# ifeq ($ (findstring _11., _ $ (shell sw_vers -productVersion)), _ 11.)
# UPLOAD_PORT? = /Dev/cu.usbserial-0001
# else
# UPLOAD_PORT? = /Dev/cu.SLAB_USBtoUART
# endif
#else
# UPLOAD_PORT? = / Dev / ttyUSB0
#endif
```
runs it fine.

Have tried ch9-poco with
blended-rectangle
works fine.

Hope you can find a solution that works for everyone.

Yours sincerely
Poul

**Comment by Frida854** [comment, marked as answer]
After some searching in makefiles, it seems that after the first time a variable is used, "? =" is used.
But if it needs to be changed afterwards, use "=".

So now I have changed "make.esp32.mk" to:
```
HOST_OS := $(shell uname)

PROJ_DIR_TEMPLATE = $(BUILD_DIR)/devices/esp32/xsProj-$(ESP32_SUBCLASS)

UPLOAD_PORT ?= $(shell bash -c "python $(PROJ_DIR_TEMPLATE)/getPort.py")

ifeq ($(HOST_OS),Darwin)
	ifeq ($(findstring _11.,_$(shell sw_vers -productVersion)),_11.)
		UPLOAD_PORT = /dev/cu.usbserial-0001
	else
		UPLOAD_PORT = /dev/cu.SLAB_USBtoUART
	endif
else
	UPLOAD_PORT = /dev/ttyUSB0
endif
```
And it works fine for me.
Yours sincerely
Poul

**Comment by andycarle** [comment]
Hi @Frida854,

Thank you for the report! 

I've done some looking into this. I concur that the logic we use in `getPort.py` needs to be updated to account for changes in serial device files on the m1 MacBooks. It looks like a related change was made recently in the ESP-IDF to account for this same problem. 

I'll take a look at good options here. In the meantime, I recommend just setting the `UPLOAD_PORT` environment variable in your shell to the path that you know works: `/dev/cu.usbserial-0001`. You can do that in your `~/.zshrc` file along with the other Moddable environment variables you've already defined. 

The change you've proposed above will also work on your machine, as you've shown, so no need to change it if you don't want to. But it may cause issues on other machines, especially those with multiple serial devices connected.

&nbsp;- Andy
